src/libopensc/card-myeid.c

```diff
@@ -1185,6 +1254,144 @@ static int myeid_decipher(struct sc_card *card, const u8 * crgram,
+		memcpy(out, apdu.resp, apdu.resplen);
+
+	LOG_FUNC_RETURN(ctx, apdu.resplen);
+}
+
+static int myeid_unwrap_key(struct sc_card *card, const u8 *crgram, size_t crgram_len)
+{
+	int r;
+	u8 out[512];
+	size_t outlen = 512;
+	struct sc_apdu apdu;
+	u8 rbuf[SC_MAX_APDU_BUFFER_SIZE];
+	u8 sbuf[SC_MAX_APDU_BUFFER_SIZE];
+	u8 p2 = 0x86;
+	myeid_private_data_t* priv;
+
+	LOG_FUNC_CALLED(card->ctx);
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_NORMAL);
+
+	assert(card != NULL && crgram != NULL && out != NULL);
+
+	if (crgram_len > 256)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+
+	if (card->drv_data)
+	{
+		priv = (myeid_private_data_t*) card->drv_data;
+
+		if (priv->sec_env)
+		{
+			if (priv->sec_env->algorithm == SC_ALGORITHM_AES ||
+				priv->sec_env->algorithm == SC_ALGORITHM_3DES ||
+				priv->sec_env->algorithm == SC_ALGORITHM_DES)
+						p2 = 0x84;		/* Set correct P2 for symmetric crypto */
+		}
+	}
+
+	/* INS: 0x2A  PERFORM SECURITY OPERATION
+	    * P1:  0x00  Do not expect response - the deciphered data will be placed into the target key EF.
+	    * P2:  0x86  Cmd: Padding indicator byte followed by cryptogram
+		* P2:  0x84  Cmd: AES/3DES Cryptogram (plain value encoded in BER-TLV DO, but not including SM DOs) */
+	sc_format_apdu(card, &apdu,
+		(crgram_len < 256) ? SC_APDU_CASE_4_SHORT : SC_APDU_CASE_3_SHORT,
+		0x2A, 0x00, p2);
+
+	apdu.resp = rbuf;
+	apdu.resplen = sizeof(rbuf);
+	apdu.le = crgram_len;
+
+	if (crgram_len == 256 && p2 == 0x86)
+	{
+		apdu.le = 0;
+		/* padding indicator byte, 0x81 = first half of 2048 bit cryptogram */
+		sbuf[0] = 0x81;
+		memcpy(sbuf + 1, crgram, crgram_len / 2);
+		apdu.lc = crgram_len / 2 + 1;
+	}
+	else if (p2 == 0x86)
+	{
+		sbuf[0] = 0; /* padding indicator byte, 0x00 = No further indication */
+		memcpy(sbuf + 1, crgram, crgram_len);
+		apdu.lc = crgram_len + 1;
+	}
+	else /* symmetric crypto, no padding indicator byte */
+	{
+		memcpy(sbuf, crgram, crgram_len);
+		apdu.lc = crgram_len;
+	}
+
+	apdu.datalen = apdu.lc;
+	apdu.data = sbuf;
+
+	r = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
+	if (apdu.sw1 == 0x90 && apdu.sw2 == 0x00)
+	{
+		if (crgram_len == 256 && p2 == 0x86) /* with symmetric crypto, we support only single apdu unwrap for now */
+		{
+			sc_format_apdu(card, &apdu, SC_APDU_CASE_4_SHORT,
+				0x2A, 0x00, p2);
+			apdu.resp = rbuf;
+			apdu.resplen = sizeof(rbuf);
+			apdu.le = crgram_len;
+			/* padding indicator byte,
+			    * 0x82 = Second half of 2048 bit cryptogram */
+			sbuf[0] = 0x82;
+			memcpy(sbuf + 1, crgram + crgram_len / 2, crgram_len / 2);
+			apdu.lc = crgram_len / 2 + 1;
+			apdu.datalen = apdu.lc;
+			apdu.data = sbuf;
+
+			r = sc_transmit_apdu(card, &apdu);
+
+			LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
+
+			if (apdu.sw1 == 0x90 && apdu.sw2 == 0x00)
+			{
+				LOG_FUNC_RETURN(card->ctx, r);
+			}
+		}
+		else
+		{
+			int len = apdu.resplen > outlen ? outlen : apdu.resplen;
```

Comment:
After all we don't need to copy data to out buffer in this function. But the idea has been to return the required buffer length to caller, so that caller can allocate long enough buffer. This convention seems to be used at least in decipher. I looked at other drivers too and they work like this. On the other hand, compute_signature returns SC_ERROR_BUFFER_TOO_SMALL if the buffer cannot hold the data returned from card. Do you know why there are different conventions in these two functions? 

I am returning required buffer size also in myeid_wrap_key. 

PKCS#11 defines returning buffer length in section 11.2: Conventions for functions returning output in a variable-length buffer
How does it work with OpenSC if you call C_Sign with pSignature = NULL?